flathunter/crawl_immobilienscout.py

```diff
@@ -88,28 +96,38 @@ def get_entries_from_javascript(self):
 
     def get_entries_from_json(self, json):
         """Get entries from JSON"""
-        jsonpath_expr = parse("$..['resultlist.realEstate']")
         return [
-          self.extract_entry_from_javascript(entry.value) for entry in jsonpath_expr.find(json)
+            self.extract_entry_from_javascript(entry.value) for entry in self.JSON_PATH_PARSER_ENTRIES.find(json)
         ]
 
     def extract_entry_from_javascript(self, entry):
         """Get single entry from JavaScript"""
-        image_path = parse("$..galleryAttachments..['@xlink.href']")
+
+        # the url that is being returned to the frontend has a placeholder for screen size. (%WIDTH% and %HEIGHT%)
+        # The website's frontend fills these variables based on the user's screen size.
+        # If we remove this part, the API will return the original size of the image.
+        #
+        # Before:
+        # https://pictures.immobilienscout24.de/listings/$$IMAGE_ID$$.jpg/ORIG/legacy_thumbnail/%WIDTH%x%HEIGHT%3E/format/webp/quality/50
+        #
+        # After: https://pictures.immobilienscout24.de/listings/$$IMAGE_ID$$.jpg
+
+        images = [
+            '/'.join(image.value.split('/')[:5]) for image in self.JSON_PATH_PARSER_IMAGES.find(entry)
+        ]
+
         return {
-            'id': int(entry["@id"]),
-            'url': ("https://www.immobilienscout24.de/expose/" + str(entry["@id"])),
-            'image': next(
-                iter([ galleryImage.value for galleryImage in image_path.find(entry) ]),
-                ("https://www.static-immobilienscout24.de/"
-                "statpic/placeholder_house/496c95154de31a357afa978cdb7f15f0_placeholder_medium.png")
-            ),
-            'title': entry["title"],
-            'address': entry["address"]["description"]["text"],
+            'id': int(entry.get("@id", 0)),
+            'url': ("https://www.immobilienscout24.de/expose/" + str(entry.get("@id", 0))),
+            'image': images[0] if len(images) else self.FALLBACK_IMAGE_URL,
+            'images': images,
+            'title': entry.get("title", ''),
+            'address': entry.get("address", {}).get("description", {}).get("text", ''),
```

Comment:
Shouldn't "address", "description", "price", "calculatedTotalRent" and "totalRent" default to empty strings, other than empty objects?

Reply:
it should be object since we're calling get after it. if we set empty string for description or address as fallback, it will throw error since string doesn't have `get` method.